src/python/grpcio/grpc/experimental/aio/_interceptor.py

```diff
@@ -0,0 +1,345 @@
+    RPC has finished but again the execution is still held by the interceptors.
+
+    Once the RPC is finally executed, all methods are finally done against the
+    intercepted call, being at the same time the same call returned to the
+    interceptors.
+
+    For most of the methods, like `initial_metadata()` the caller does not need
+    to wait until the interceptors task is finished, once the RPC is done the
+    caller will have the freedom for accessing to the results.
+
+    For the `__await__` method is it is proxied to the intercepted call only when
+    the interceptor task is finished.
+    """
+
+    _loop: asyncio.AbstractEventLoop
+    _channel: cygrpc.AioChannel
+    _cancelled_before_rpc: bool
+    _intercepted_call: Optional[_base_call.UnaryUnaryCall]
+    _intercepted_call_created: asyncio.Event
+    _interceptors_task: asyncio.Task
+
+    def __init__( # pylint: disable=R0913
+            self,
+            interceptors: Sequence[UnaryUnaryClientInterceptor],
+            request: RequestType,
+            timeout: Optional[float],
+            channel: cygrpc.AioChannel,
+            method: bytes,
+            request_serializer: SerializingFunction,
+            response_deserializer: DeserializingFunction) -> None:
+        self._channel = channel
+        self._loop = asyncio.get_event_loop()
+        self._interceptors_task = asyncio.ensure_future(
+            self._invoke(interceptors, method, timeout, request,
+                         request_serializer, response_deserializer))
+
+    def __del__(self):
+        self.cancel()
+
+    async def _invoke(
+            self, interceptors: Sequence[UnaryUnaryClientInterceptor],
+            method: bytes, timeout: Optional[float], request: RequestType,
+            request_serializer: SerializingFunction,
+            response_deserializer: DeserializingFunction) -> UnaryUnaryCall:
+        """Run the RPC call wrapped in interceptors"""
+
+        async def _run_interceptor(
+                interceptors: Iterator[UnaryUnaryClientInterceptor],
+                client_call_details: ClientCallDetails,
+                request: RequestType) -> _base_call.UnaryUnaryCall:
+            try:
+                interceptor = next(interceptors)
+            except StopIteration:
+                interceptor = None
+
+            if interceptor:
+                continuation = functools.partial(_run_interceptor, interceptors)
+                try:
+                    call_or_response = await interceptor.intercept_unary_unary(
+                        continuation, client_call_details, request)
+                except grpc.RpcError as err:
+                    # gRPC error is masked inside an artificial call,
+                    # caller will see this error if and only
+                    # if runs an `await call` operation
+                    return UnaryUnaryCallRpcError(err)
+                except asyncio.CancelledError:
```

Comment:
@lidizheng this is one of the main pain points having to deal with local cancellations masked as simple and generic `asyncio.CancelledError`.

For making them available to the previous interceptor or to the caller implementing the following flow:

```python
call = multicallable(...)
try:
    await call
except asyncio.CancelledError
    print(await call.code()) 
```
We have been forced to create an artificial `UnaryUnaryCancelledError` which masks the `asyncio.CancelledError`, which IMO is debatable.

Now this artificial `UnaryUnaryCancelledError` class makes its best effort and just returns alwayS
the `grpc.StatusCode.CANCELLED` status code, which might be totally false since the RPC could not even be started.

Reply:
The solution you have is similar to what sync stack have (see [code](https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/_interceptor.py#L150). They have a `FailureOutcome` class and a `UnaryOutcome` class.

However, about the `UnaryUnaryCancelledError`, I have several thoughts to discuss:
1. If `asyncio.CancelledError` is explicitly injected by the application, should we abort the interception?
2. Should we give users the power to decide whether to abort or not? Can they use `asyncio.shield` to stop the cancellation propagation?
3. Is there a reason `UnaryUnaryCancelledError` can not merge with `UnaryUnaryCallRpcError`?

Reply:
One comment about your snippet, for completing the RPC you would need to await the call that has been returned by the continuation. Otherwise, you won't get any response or exception

Regarding the questions that you made

>If asyncio.CancelledError is explicitly injected by the application, should we abort the interception?

You mean but having another part of the app canceling with the Asyncio interface a task that might be executing an RPC? In that case, the cancellation path should be the same. I could add a new test for that scenario.
 
> Should we give users the power to decide whether to abort or not? They can use following script to stop asyncio.CancelledError bottom-up propagation.

If the user wants to swallow the local cancellation and try it again has total freedom on makeing a new call, for example:


```python
async def intercept_unary_unary(self, continuation, client_call_details, request):
    for i in range(ATTEMPTS):
        try:
            call = await continuation(client_call_details, request)
            await call
            break
        except asyncio.CancelledError:
            print('RPC cancelled locally.')
            print('Trying again')
        except grpc.RpcError as rpc_error:
            if rpc_error.code() == grpc.StatusCode.CANCELLED:
                print('RPC cancelled remotely.')
           print('Trying again')
    return call
```

> Is there a reason UnaryUnaryCancelledError can not merge with UnaryUnaryCallRpcError

Just an implementation detail, while in the `UnaryUnaryCallRpcError` RPC attributes depend on the RPC error, for the `UnaryUnaryCancelledError` everything is already pre-backed in the class.

Reply:
Merging with the comment above: https://github.com/grpc/grpc/pull/21455/files#r359028226

I am wrong. No matter the RPC is cancelled locally or remotely, the only way to "recover" from this error is retrying, there is no way to resume execution. Under this statement, I'm asking myself that is splitting `CancelledError` and `RpcError` really worth the effort? Should we merge them and provide a single-unified error handling path?

WDYT? Let's discuss in the thread above.